File: 2019/src/main/java/frc/robot/Robot.java
package frc.robot;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.command.Command;
import edu.wpi.first.wpilibj.command.Scheduler;
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.robot.commands.DrivePath;
import frc.robot.subsystems.Drivetrain;
import frc.robot.subsystems.Gyro;
import frc.robot.subsystems.Lights;

/**
 * The VM is configured to automatically run this class, and to call the
 * functions corresponding to each mode, as described in the TimedRobot
 * documentation. If you change the name of this class or the package after
 * creating this project, you must also update the build.gradle file in the
 * project.
 */
public class Robot extends TimedRobot {
	public static Gyro m_gyro = new Gyro();
	public static Lights m_lights = null; // new Lights();
	public static Drivetrain m_drivetrain = new Drivetrain();
	public static OI m_oi;

	private SendableChooser<String> m_chooserPath = new SendableChooser<>();
	private Command m_autonomousCommand;

	@Override
	public void robotInit() {
		m_oi = new OI();
		m_chooserPath.setDefaultOption("None", null);
		m_chooserPath.addOption("Left Hatch 1", "LeftHatch1.xml");
		m_chooserPath.addOption("Left Hatch 2", "LeftHatch2.xml");
		m_chooserPath.addOption("Left Hatch 3", "LeftHatch3.xml");
		SmartDashboard.putData("Auto Path", m_chooserPath);
	}

	@Override
	public void robotPeriodic() {
	}

	@Override
	public void disabledInit() {
	}

	@Override
	public void disabledPeriodic() {
		Scheduler.getInstance().run();
	}

	@Override
	public void autonomousInit() {
		String filename = m_chooserPath.getSelected();

		if (filename != null) {
			m_autonomousCommand = new DrivePath(filename);
			m_autonomousCommand.start();
		}
	}

	@Override
	public void autonomousPeriodic() {
		Scheduler.getInstance().run();
	}

	@Override
	public void teleopInit() {
		// This makes sure that the autonomous stops running when teleop starts running. 
		if (m_autonomousCommand != null) {
			m_autonomousCommand.cancel();
		}
	}

	@Override
	public void teleopPeriodic() {
		Scheduler.getInstance().run();
	}

	@Override
	public void testPeriodic() {
	}
}

File: 2019_cpp/src/main/cpp/Robot.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.												*/
/* Open Source Software - may be modified and shared by FRC teams. The code	 */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.																															 */
/*----------------------------------------------------------------------------*/

#include "Robot.h"

#include <frc/commands/Scheduler.h>
#include <frc/smartdashboard/SmartDashboard.h>

Gyro Robot::m_gyro = Gyro();
Lights Robot::m_lights = Lights();
OI Robot::m_oi = OI();

void Robot::RobotInit()
{
	m_chooser.SetDefaultOption("None", "");
	m_chooser.AddOption("Left Hatch 1", "LeftHatch1.xml");
	m_chooser.AddOption("Left Hatch 2", "LeftHatch2.xml");
	m_chooser.AddOption("Left Hatch 3", "LeftHatch3.xml");
	frc::SmartDashboard::PutData("Auto Path", &m_chooser);
}

void Robot::RobotPeriodic()
{
}

void Robot::DisabledInit()
{
}

void Robot::DisabledPeriodic()
{ 
	frc::Scheduler::GetInstance()->Run(); 
}

void Robot::AutonomousInit()
{
	std::string filename = m_chooser.GetSelected();

	if (!filename.empty())
	{
	//	m_command = std::make_unique<DrivePath>(filename);
		m_command->Start();
	}
}

void Robot::AutonomousPeriodic()
{ 
	frc::Scheduler::GetInstance()->Run(); 
}

void Robot::TeleopInit()
{
	// This makes sure that the autonomous stops running when teleop starts running. 
	if (m_command != nullptr)
	{
		m_command->Cancel();
		m_command = nullptr;
	}
}

void Robot::TeleopPeriodic()
{ 
	frc::Scheduler::GetInstance()->Run(); 
}

void Robot::TestPeriodic()
{
}

#ifndef RUNNING_FRC_TESTS
int main() { return frc::StartRobot<Robot>(); }
#endif
